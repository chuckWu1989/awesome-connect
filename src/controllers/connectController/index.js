import lodash from 'lodash';
import { connect as reduxConnect } from 'react-redux';
import enhancerRouter from '../../routers/enhancerRouter';

export const getMapStateToProps = (mapStateToProps, defaultProps) => (
  typeof mapStateToProps === 'function' ?
    (state, props) => (
      mapStateToProps(state, lodash.merge({}, defaultProps, props))
    ) :
    null
);
export const getMapDispatchToProps = (mapDispatchToProps, defaultProps) => (
  typeof mapDispatchToProps === 'function' ?
    (dispatch, props) => (
      mapDispatchToProps(dispatch, lodash.merge({}, defaultProps, props))
    ) :
    null
);
export const getMergeProps = (mergeProps, defaultProps) => (
  typeof mergeProps === 'function' ?
    (stateProps, dispatchProps, props) => (
      mergeProps(
        stateProps, dispatchProps, lodash.merge({}, defaultProps, props),
      )
    ) :
    null
);
export const connect = store => (
  (mapStateToProps = null, mapDispatchToProps = null, mergeProps = null) => (
    (component, type = {}) => {
      if (typeof component !== 'function') {
        const error = new TypeError('The component must be a function', 'connect');
        if (process.env.NODE_ENV !== 'production') {
          console.error(error.toString());
        }
        throw error;
      }
      const { defaultProps = {} } = component;
      const newMapStateToProps = getMapStateToProps(mapStateToProps, defaultProps);
      const newMapDispatchToProps = getMapDispatchToProps(mapDispatchToProps, defaultProps);
      const newMergeProps = getMergeProps(mergeProps, defaultProps);
      const ConnectComponent = reduxConnect(
        newMapStateToProps,
        newMapDispatchToProps,
        newMergeProps,
      )(component);
      const HOC = enhancerRouter(ConnectComponent, store);
      Object.keys(type).forEach((key) => {
        HOC[key] = type[key];
      });
      return HOC;
    }
  )
);
