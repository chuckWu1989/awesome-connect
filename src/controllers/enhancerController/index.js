/* eslint no-underscore-dangle: off */
import uuidv4 from 'uuid/v4';

export const getId = () => uuidv4();
export const getName = (id, name) => {
  const result = name !== undefined ? name : id;
  return result;
};
export function mergeProps(props, name, store) {
  return (
    '_reactInternalInstance' in this &&
    '_context' in this._reactInternalInstance &&
    typeof this._reactInternalInstance._context === 'object' &&
    !('store' in this._reactInternalInstance._context) ?
      { ...props, name, store } :
      { ...props, name }
  );
}
