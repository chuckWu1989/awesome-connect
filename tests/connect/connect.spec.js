import React from 'react';
import toJson from 'enzyme-to-json';
import * as controller from '../../src/controllers/connectController';

const { shallow } = global;

describe('connect', () => {
  describe('getMapStateToProps', () => {
    it('should return function if mapStateToProps is a function', () => {
      const defaultProps = {};
      const mapStateToProps = () => {};
      const result = controller.getMapStateToProps(mapStateToProps, defaultProps);
      expect(typeof result).toBe('function');
    });
    it('should return null if mapStateToProps is not a function', () => {
      const defaultProps = {};
      const mapStateToProps = undefined;
      const result = controller.getMapStateToProps(mapStateToProps, defaultProps);
      expect(result).toBe(null);
    });
    it('should get merged props if mapStateToProps is a function', () => {
      let result;
      const defaultProps = { default: 'test' };
      const props = { newProps: 'test2' };
      const mapStateToProps = (state, ownProps) => { result = ownProps; };
      controller.getMapStateToProps(mapStateToProps, defaultProps)(null, props);
      expect('default' in result && 'newProps' in result).toBeTruthy();
    });
  });
  describe('getMapDispatchToProps', () => {
    it('should return function if mapDispatchToProps is a function', () => {
      const defaultProps = {};
      const mapDispatchToProps = () => {};
      const result = controller.getMapDispatchToProps(mapDispatchToProps, defaultProps);
      expect(typeof result).toBe('function');
    });
    it('should return null if mapDispatchToProps is not a function', () => {
      const defaultProps = {};
      const mapDispatchToProps = undefined;
      const result = controller.getMapDispatchToProps(mapDispatchToProps, defaultProps);
      expect(result).toBe(null);
    });
    it('should get merged props if mapDispatchToProps is a function', () => {
      let result;
      const defaultProps = { default: 'test' };
      const props = { newProps: 'test2' };
      const mapDispatchToProps = (dispatch, ownProps) => { result = ownProps; };
      controller.getMapDispatchToProps(mapDispatchToProps, defaultProps)(null, props);
      expect('default' in result && 'newProps' in result).toBeTruthy();
    });
  });
  describe('getMergeProps', () => {
    it('should return function if mergeProps is a function', () => {
      const defaultProps = {};
      const mergeProps = () => {};
      const result = controller.getMergeProps(mergeProps, defaultProps);
      expect(typeof result).toBe('function');
    });
    it('should return null if mergeProps is not a function', () => {
      const defaultProps = {};
      const mergeProps = undefined;
      const result = controller.getMergeProps(mergeProps, defaultProps);
      expect(result).toBe(null);
    });
    it('should get merged props if mergeProps is a function', () => {
      let result;
      const defaultProps = { default: 'test' };
      const props = { newProps: 'test2' };
      const mergeProps = (stateProps, dispatchProps, ownProps) => { result = ownProps; };
      controller.getMergeProps(mergeProps, defaultProps)(null, null, props);
      expect('default' in result && 'newProps' in result).toBeTruthy();
    });
  });
  describe('connect', () => {
    it('should return function by calling once', () => {
      const result = controller.connect();
      expect(typeof result).toBe('function');
    });
    it('should return function by calling twice', () => {
      const result = controller.connect()();
      expect(typeof result).toBe('function');
    });
    it('should output console error if component is not a function and env is not production', () => {
      let err;
      console.error = jest.fn();
      try {
        controller.connect()()();
      } catch (e) {
        err = e;
      }
      expect(err).toBeDefined();
      expect(console.error).toHaveBeenCalledTimes(1);
    });
    it('should throw error only if component is not a function and env is production', () => {
      let err;
      process.env.NODE_ENV = 'production';
      console.error = jest.fn();
      try {
        controller.connect()()();
      } catch (e) {
        err = e;
      }
      process.env.NODE_ENV = 'test';
      expect(err).toBeDefined();
      expect(console.error).toHaveBeenCalledTimes(0);
    });
    it('should return a connected component if component is a function', () => {
      const component = () => <div />;
      component.defaultProps = { default: 'test' };
      const ConnectComponent = controller.connect()()(component);
      const tree = toJson(shallow(<ConnectComponent />));
      expect(tree).toMatchSnapshot();
    });
    it('should include type if type is defined', () => {
      const component = () => <div />;
      const type = { default: 'test', test: 'test2' };
      const ConnectComponent = controller.connect()()(component, type);
      expect('default' in ConnectComponent && 'test' in ConnectComponent).toBeTruthy();
    });
  });
});
