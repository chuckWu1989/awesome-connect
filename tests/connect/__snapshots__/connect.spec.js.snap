// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connect connect should return a connected component if component is a function 1`] = `
<_class
  getId={[Function]}
  getName={[Function]}
  mergeProps={[Function]}
/>
`;
